lib/Transforms/LLZKR1CSLoweringPass.cpp

```diff
@@ -0,0 +1,314 @@
+  /// only when strictly necessary. This function ensures the resulting expression:
+  ///
+  /// - Has at most one multiplication per constraint (R1CS-compatible)
+  /// - Avoids unnecessary introduction of auxiliary variables
+  /// - Preserves semantic equivalence via auxiliary field equality constraints
+  ///
+  /// Rewriting is done **bottom-up** using post-order traversal of the def-use chain.
+  /// The transformation is minimal:
+  /// - Only rewrites Add/Sub where both operands are degree-2
+  /// - Leaves multiplications intact unless their operands require rewriting due to constants
+  /// - Avoids rewriting expressions that are already linear or already normalized
+  ///
+  /// The function memoizes all degrees and rewrites for efficiency and correctness,
+  /// and records any auxiliary field assignments for later reconstruction in compute().
+  ///
+  /// \param root           The root felt-valued expression to normalize.
+  /// \param structDef      The enclosing struct definition (for adding aux fields).
+  /// \param constrainFunc  The constrain() function containing the constraint logic.
+  /// \param degreeMemo     Memoized degrees of expressions (to avoid recomputation).
+  /// \param rewrites       Memoized rewrites of expressions.
+  /// \param auxAssignments Records auxiliary field assignments introduced during normalization.
+  /// \param builder        Builder used to insert new ops in the constrain() block.
+  /// \returns              A Value representing the normalized (possibly rewritten) expression.
+  Value normalizeForR1CS(
+      Value root, StructDefOp structDef, FuncDefOp constrainFunc,
+      DenseMap<Value, unsigned> &degreeMemo, DenseMap<Value, Value> &rewrites,
+      SmallVectorImpl<AuxAssignment> &auxAssignments, OpBuilder &builder
+  ) {
+    if (rewrites.count(root)) {
+      return rewrites[root];
+    }
+
+    SmallVector<Value, 16> postOrder;
+    getPostOrder(root, postOrder);
+
+    // We perform a bottom up rewrite of the expressions. For any expression e := op(e_1, ...,
+    // e_n) we first rewrite e_1, ..., e_n if necessary and then rewrite e based on op.
+    for (Value val : postOrder) {
+      if (rewrites.count(val)) {
+        continue;
+      }
+
+      Operation *op = val.getDefiningOp();
+
+      if (!op) {
+        // Block arguments, etc.
+        degreeMemo[val] = 1;
+        rewrites[val] = val;
+        continue;
+      }
+
+      // Case 1: Felt constant op. The degree is 0 and no rewrite is needed.
+      if (auto c = val.getDefiningOp<FeltConstantOp>()) {
+        degreeMemo[val] = 0;
+        rewrites[val] = val;
+        continue;
+      }
+
+      // Case 2: Field read op. The degree is 1 and no rewrite needed.
+      if (auto fr = val.getDefiningOp<FieldReadOp>()) {
+        degreeMemo[val] = 1;
+        rewrites[val] = val;
+        continue;
+      }
+
+      // Helper function for getting degree from memo map
+      auto getDeg = [&](Value v) -> unsigned {
```

Comment:
Prefer minimal captures when feasible
```suggestion
      auto getDeg = [&degreeMemo](Value v) -> unsigned {
```